File: src/main/java/dev/slne/discord/discord/interaction/modal/step/creator/whitelist/step/WhitelistTicketMinecraftNameStep.java
package dev.slne.discord.discord.interaction.modal.step.creator.whitelist.step;

import dev.slne.data.api.DataApi;
import dev.slne.discord.Bootstrap;
import dev.slne.discord.discord.interaction.modal.step.MessageQueue;
import dev.slne.discord.discord.interaction.modal.step.ModalComponentBuilder;
import dev.slne.discord.discord.interaction.modal.step.ModalStep;
import dev.slne.discord.message.RawMessages;
import dev.slne.discord.spring.service.whitelist.WhitelistService;
import java.util.UUID;
import java.util.concurrent.CompletionException;
import net.dv8tion.jda.api.entities.channel.concrete.TextChannel;
import net.dv8tion.jda.api.events.interaction.ModalInteractionEvent;
import net.dv8tion.jda.api.interactions.components.text.TextInput;
import net.dv8tion.jda.api.interactions.components.text.TextInputStyle;
import org.springframework.data.util.Lazy;

public class WhitelistTicketMinecraftNameStep extends ModalStep {

  private static final Lazy<WhitelistService> WHITELIST_SERVICE = Lazy.of(
      () -> Bootstrap.getContext().getBean(
          WhitelistService.class));
  private static final String MINECRAFT_NAME = "minecraft-name";

  private String minecraftName;

  public WhitelistTicketMinecraftNameStep(ModalStep parent) {
  }

  @Override
  protected void buildModalComponents(ModalComponentBuilder builder) {
    builder.addComponent(
        TextInput.create(
                MINECRAFT_NAME,
                RawMessages.get("modal.whitelist.step.minecraft.label"),
                TextInputStyle.SHORT
            )
            .setRequired(true)
            .setRequiredRange(3, 16)
            .build()
    );
  }

  @Override
  protected void verifyModalInput(ModalInteractionEvent event)
      throws ModalStepInputVerificationException {
    this.minecraftName = getRequiredInput(event, MINECRAFT_NAME);
    final UUID uuid;

    try {
      uuid = DataApi.getUuidByPlayerName(this.minecraftName).join();
    } catch (CompletionException e) {
      throw new ModalStepInputVerificationException(
          RawMessages.get("modal.whitelist.step.minecraft.invalid"), e);
    }

    if (WHITELIST_SERVICE.get().isWhitelisted(uuid, null, null).join()) {
      throw new ModalStepInputVerificationException(
          RawMessages.get("interaction.command.ticket.whitelist.already-whitelisted"));
    }
  }

  @Override
  protected void buildOpenMessages(MessageQueue messages, TextChannel channel) {
    messages.addMessage(RawMessages.get("modal.whitelist.step.minecraft.open", this.minecraftName));
  }
}

File: src/main/kotlin/dev/slne/discord/discord/interaction/modal/step/creator/whitelist/step/WhitelistTicketMinecraftNameStep.kt
package dev.slne.discord.discord.interaction.modal.step.creator.whitelist.step

import dev.slne.discord.discord.interaction.modal.step.MessageQueue
import dev.slne.discord.discord.interaction.modal.step.ModalStep
import dev.slne.discord.message.translatable
import dev.slne.discord.persistence.service.user.UserService
import dev.slne.discord.persistence.service.whitelist.WhitelistService
import net.dv8tion.jda.api.components.label.Label
import net.dv8tion.jda.api.components.textinput.TextInput
import net.dv8tion.jda.api.components.textinput.TextInputStyle
import net.dv8tion.jda.api.entities.channel.concrete.ThreadChannel
import net.dv8tion.jda.api.events.interaction.ModalInteractionEvent
import net.dv8tion.jda.api.interactions.components.selections.StringSelectInteraction
import net.dv8tion.jda.api.modals.Modal

private const val MINECRAFT_NAME = "minecraft-name"

class WhitelistTicketMinecraftNameStep(
    parent: ModalStep,
    private val userService: UserService,
    private val whitelistService: WhitelistService
) : ModalStep() {

    private var minecraftName: String? = null

    override fun buildModalComponents(builder: Modal.Builder, interaction: StringSelectInteraction) {
        builder.addComponents(
            Label.of(
                translatable("modal.whitelist.step.minecraft.label"),
                TextInput.create(MINECRAFT_NAME, TextInputStyle.SHORT)
                    .setRequired(true)
                    .setRequiredRange(3, 16)
                    .build()
            )
        )
    }

    override suspend fun verifyModalInput(event: ModalInteractionEvent) {
        minecraftName = getInput(event, MINECRAFT_NAME)

        val uuid = userService.getUuidByUsername(minecraftName!!)
            ?: throw ModalStepInputVerificationException(
                translatable("modal.whitelist.step.minecraft.invalid")
            )

        if (whitelistService.isWhitelisted(uuid, null, null)) {
            throw ModalStepInputVerificationException(
                translatable("interaction.command.ticket.whitelist.already-whitelisted")
            )
        }
    }

    override fun MessageQueue.buildOpenMessages(thread: ThreadChannel) {
        addMessage(
            translatable(
                "modal.whitelist.step.minecraft.open",
                minecraftName
            )
        )
    }
}
